### PR TITLE
sql: scan the system.role_members table using a single scan

### LINKED ISSUE #77452: sql: Improve role membership cold cache scenarios
This issue tracks two problems:

In order to check privileges for SQL commands, we must check not just the privilege for the current user, but also all parent (and grandparent, etc) roles for that user.

Normally, the role membership information is cached, so checking privileges is quick. However, this cache of memberships needs to be cleared whenever users are modified.

We have received reports of slowness (~10 seconds) coming from the time it takes to rebuild the cache. Rebuilding the cache can happen WHILE executing almost any command, since it happens while checking if the user has privileges to run the command.

We've found two problems:

1. We have some inefficient logic that means we run multiple internal queries to construct the role membership tree. Specifically, we do one query for each layer of the role membership hierarchy. This is bad because each of these queries can result in a network hop to another node in the cluster. we can change this so it happens all in one query. See https://github.com/cockroachdb/cockroach/blob/6880665c589c6a49e3f985ed2cc3448052d26763/pkg/sql/authorization.go#L467-L516
2. Another problem is that if there are multiple user-initiated queries happening at the same time, then it’s possible for each one of them to try to repopulate the cache. This means each one of them will kick off multiple internal queries, causing more overhead in the cluster. We can fix this so that there’s only one “cache reconstruction” happening at a time. See https://github.com/cockroachdb/cockroach/blob/6880665c589c6a49e3f985ed2cc3448052d26763/pkg/sql/authorization.go#L430-L433

This issue tracks improving those cases. We currently are planning to create a backport for problem 1, but not problem 2.

Jira issue: CRDB-13599
Epic CRDB-8948

### PR DESCRIPTION
**Pull request by rafiss**
relates to support case https://github.com/cockroachlabs/support/issues/1463
fixes https://github.com/cockroachdb/cockroach/issues/77452

### sql: use singleflight for role membership lookups

This helps prevent a thundering herd when this cache gets invalidated.

### sql: scan the system.role_members table using a single scan

Release note (sql change): Added a
`sql.auth.resolve_membership_single_scan.enabled` cluster setting, which
changes the query for an internal role membership cache. Previously the
code would recursively look up each role in the membership hierarchy,
leading to multiple queries. With the setting on, it uses a single
query. The setting is true by default.

### sql: fix information_schema role introspection

Release note (bug fix): The information_schema tables
administrable_role_authorizations and applicable_roles were incorrectly
always returning the current user for the grantee column. Now, the
column will contain the correct role that was granted the parent role
given in the role_name column.

Release justification: high priority bug fix

### CODE REVIEW

File: pkg/sql/user_test.go [hunk 1]
```diff
@@ -48,190 +49,199 @@ func TestGetUserTimeout(t *testing.T) {
...
-			dbSQL, err := pgxConn(t, rootURL)
-			if err != nil {
-				t.Fatal(err)
-			}
-			defer func() { _ = dbSQL.Close(ctx) }()
-			// A simple query must work for 'root' even without a system range available.
-			if _, err := dbSQL.Exec(ctx, "SELECT 1"); err != nil {
-				t.Fatal(err)
+			if singleScan {
+				_, err := db.Exec(`SET CLUSTER SETTING sql.auth.resolve_member_single_scan.enabled = true`)
```
Comment (otan) [inline pkg/sql/user_test.go:85]: Maybe take out the if statement

**Review by otan — COMMENTED** [review]
Quick query: are we planning to backport this?

**Comment by rafiss** [comment]
yeah, i was planning to backport, but change the cluster setting to default to false in 21.2

File: pkg/sql/user_test.go (hunk 1 above)
Comment (rafiss) [reply pkg/sql/user_test.go]: done

**Review by ajwerner — COMMENTED** [review]
> Reviewed 1 of 4 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77359)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @otan, @rafiss, and @RichardJCai)

___
*[pkg/sql/authorization.go, line 497 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/77359#-MxXgAu42pKE5Q-fc3KW:-MxXgAu42pKE5Q-fc3KX:b-frzccn) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a1e611b69650706917541c44492e77fdb752cc0/pkg/sql/authorization.go#L497)):*
> ```Go
>         INNER JOIN cte_members AS child ON parent.member = child.role
>     )
> SELECT * FROM cte_members`
> ```

is the plan here actually good? Is this really a good idea? On some level this seems great, but on another, we've already committed to keeping the whole cache in RAM. :shrug: 



_Code quote:_
```Go

  cte_members (member, role, "isAdmin")
    AS (
      SELECT member, role, "isAdmin"
      FROM system.role_members
      WHERE member = $1
      UNION ALL
        SELECT parent.member, parent.role, parent."isAdmin"
        FROM system.role_members AS parent
        INNER JOIN cte_members AS child ON parent.member = child.role
    )
SELECT * FROM cte_members`
```
___
*[pkg/sql/authorization.go, line 498 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/77359#-MxXgUo04sw6oIxnYCJR:-MxXgUo04sw6oIxnYCJS:b4h0yv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a1e611b69650706917541c44492e77fdb752cc0/pkg/sql/authorization.go#L498)):*
> ```Go
>     )
> SELECT * FROM cte_members`
> 		it, err := ie.QueryIterator(
> ```

does this need a defer `it.Close`? I can see the argument why it doesn't, namely there's not other early error return, but I'd rather you put it because of panics and future code changes. 
___
*[pkg/sql/information_schema.go, line 2675 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77359#-MxXgxQdE0lvFgg96V88:-MxXgxQdE0lvFgg96V89:b-bbacik) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8400cab8b0620cd8f7c70c75a47427d6cc2dca/pkg/sql/information_schema.go#L2675)):*
> ```Go
> }
> 
> func forEachRoleHierarchyOfUser(
> ```

it feels like we should factor this out

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77359)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @otan, and @RichardJCai)

___
*[pkg/sql/authorization.go, line 497 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/77359#-MxXgAu42pKE5Q-fc3KW:-Mx_L4LR82Pj-4TX8L3W:b-a2rerh) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a1e611b69650706917541c44492e77fdb752cc0/pkg/sql/authorization.go#L497)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

is the plan here actually good? Is this really a good idea? On some level this seems great, but on another, we've already committed to keeping the whole cache in RAM. :shrug: 
</blockquote></details>

I'm experimenting a bit with sample data to see what works best
___
*[pkg/sql/authorization.go, line 498 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/77359#-MxXgUo04sw6oIxnYCJR:-Mx_PcQJEeZH3tLrQ94v:bg796yz) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a1e611b69650706917541c44492e77fdb752cc0/pkg/sql/authorization.go#L498)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

does this need a defer `it.Close`? I can see the argument why it doesn't, namely there's not other early error return, but I'd rather you put it because of panics and future code changes. 
</blockquote></details>

makes sense, adding.
___
*[pkg/sql/information_schema.go, line 2675 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77359#-MxXgxQdE0lvFgg96V88:-Mx_Peaw-f2Up0oFmiIs:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8400cab8b0620cd8f7c70c75a47427d6cc2dca/pkg/sql/information_schema.go#L2675)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

it feels like we should factor this out 
</blockquote></details>

done

File: pkg/sql/authorization.go [hunk 2]
```diff
@@ -464,14 +465,58 @@ func MemberOfWithAdminOption(
...
+	ctx context.Context,
+	member security.SQLUsername,
+	ie sqlutil.InternalExecutor,
+	txn *kv.Txn,
+	singleQuery bool,
 ) (map[security.SQLUsername]bool, error) {
 	ret := map[security.SQLUsername]bool{}
+	if singleQuery {
+		const lookupRolesStmt = `
+WITH RECURSIVE
```
Comment (otan) [inline pkg/sql/authorization.go:515]: this is nice! i wonder if we should test false cases more though by, e.g. running all of the `role` logic tests with true and false (maybe by defining a new logic test setup like `5node`)

**Review by otan — APPROVED** [review]
looking at first and third commits only (second is a separate PR i believe)
i would appreciate that extra test mentioned

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77359)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @otan, @rafiss, and @RichardJCai)

___
*[pkg/sql/authorization.go, line 497 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/77359#-MxXgAu42pKE5Q-fc3KW:-MxaR9A24WjRIygRPGpA:b-1cn08) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a1e611b69650706917541c44492e77fdb752cc0/pkg/sql/authorization.go#L497)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

I'm experimenting a bit with sample data to see what works best
</blockquote></details>

I did a few benchmarks. I created 150 users and randomly added 500 memberships between them. That's not super realistic, since it leads to a role hierarchy with many levels, but it showed that the performance of recursive CTEs can quickly get out of control.

Also, it helped me see that my first attempt at a recursive CTE has a bug, since `UNION ALL` would lead to duplicate rows being included in the output. Adding a `DISTINCT` in the outer SELECT query fixes it, but that makes performance even worse. Using `UNION` (without all) fixes the duplicates bug and performs better than `UNION ALL`, but it's only available in v22.1.

```
BenchmarkLotsOfRoles/full_scan         	     338	   3305653 ns/op

BenchmarkLotsOfRoles/recursive_cte_union         	     157	   7131301 ns/op

BenchmarkLotsOfRoles/recursive_cte_union_all    did not complete
```

**Comment by ajwerner** [comment]
@rafiss can you try out this query on your data:

```sql
WITH RECURSIVE cte_members (role, path, admin, cycle) AS (
                         SELECT role, ARRAY[$1], "isAdmin", false
                           FROM system.role_members
                          WHERE member = $1
                         UNION ALL SELECT parent.role,
                                          child.path || parent.member,
                                          parent."isAdmin",
                                          parent.member = ANY (child.path)
                                     FROM system.role_members AS parent
                               INNER JOIN cte_members AS child ON parent.member
                                                                  = child.role
                                                              AND NOT cycle
                     )
        SELECT $1, role, bool_or(admin) AS "isAdmin" -- note we don't need to return the member back
          FROM cte_members
      GROUP BY role;
```

**Comment by rafiss** [comment]
That query also is taking a long time (it's been running more than 1 minute).

**Comment by otan** [comment]
should we just use a single SQL scan then?

**Comment by rafiss** [comment]
> should we just use a single SQL scan then?

yep, done

File: pkg/sql/authorization.go [hunk 3]
```diff
@@ -428,10 +436,31 @@ func MemberOfWithAdminOption(
...
+		)
+		var memberships map[security.SQLUsername]bool
+		select {
+		case res := <-ch:
+			if res.Err != nil {
+				return nil, res.Err
+			}
+			memberships = res.Val.(map[security.SQLUsername]bool)
+		case <-ctx.Done():
+			return nil, ctx.Err()
```
Comment (otan) [inline pkg/sql/authorization.go:463]: remind me - should we wrap this error?

File: pkg/sql/logictest/testdata/logic_test/role [hunk 4]
```diff
@@ -181,18 +181,33 @@ user root
 statement ok
```
Comment (otan) [inline pkg/sql/logictest/testdata/logic_test/role:181]: can we run this logic test file with cluster setting on and off?
add something to https://github.com/cockroachdb/cockroach/blob/c6a5a395cbecbbd5ba77c2c14e3bf68eb59eeb19/pkg/sql/logictest/logic.go#L679-L684 and add `LogicTest: local .... local-cache-single-read-off` or something?

**Review by otan — APPROVED** [review]
(lgtm otherwise but i'd love that test)

**Comment by rafiss** [comment]
I changed the default value of `sql.auth.resolve_membership_single_scan.enabled` to be a metamorphic constant, so test builds have a chance of defaulting to false.

File: pkg/sql/logictest/testdata/logic_test/role (hunk 4 above)
Comment (rafiss) [reply pkg/sql/logictest/testdata/logic_test/role:181]: added a metamorphic constant

**Comment by rafiss** [comment]
bors r=otan

**Pull request merged**